bson/src/main/org/bson/codecs/pojo/PojoCodecImpl.java

```diff
@@ -347,7 +255,10 @@ private void decodeProperties(final BsonReader reader, final DecoderContext deco
                 if (discriminatorKey.equals(name)) {
                     discriminatorKeyFound = true;
                     try {
-                        codec = (Codec<T>) registry.get(discriminatorLookup.lookup(reader.readString()));
+                        Class<?> discriminatorClass = discriminatorLookup.lookup(reader.readString());
+                        if (!codec.getEncoderClass().equals(discriminatorClass)) {
+                            codec = (Codec<T>) registry.get(discriminatorClass);
```

Comment:
Only go to the codec registry if needed. Previously this worked because the codec registry was changed to be; `fromRegistries(fromCodecs(this), codecRegistry);` in the constructor.